src/utilities/routes.ts

```diff
@@ -1,5 +1,36 @@
-import { MaybeDeepReadonly, ResolvedRoute, RouteMiddleware, isNamedRoute } from '@/types'
+import { ResolvedRoute, RouteMiddleware, isNamedRoute } from '@/types'
 import { asArray } from '@/utilities/array'
+import { RouterRoute } from '@/utilities/createRouterRoute'
+import { createRouterRouteQuery } from '@/utilities/createRouterRouteQuery'
+import { getRouteParams, routeParamsAreValid } from '@/utilities/paramValidation'
+import { routePathMatches, routeQueryMatches } from '@/utilities/routeMatchRegexRules'
+import { getRouteScoreSortMethod } from '@/utilities/routeMatchScore'
+
+export function getRouterRouteForUrl(routes: ResolvedRoute[], url: string): RouterRoute | undefined {
+  const rules = [routePathMatches, routeQueryMatches, routeParamsAreValid]
+  const sortByRouteScore = getRouteScoreSortMethod(url)
+
+  const matches = routes
+    .filter(route => rules.every(test => test(route, url)))
+    .sort(sortByRouteScore)
+
+  if (matches.length === 0) {
+    return undefined
+  }
+
+  const [route] = matches
+  const [, search] = url.split('?')
```

Comment:
I think using browser `URL` is probably preferred here?

```suggestion
  const { search } = createMaybeRelativeUrl(url)
```

Reply:
Doesn't that require a full url with a domain?

Reply:
that's why I suggested using `createMaybeRelativeUrl`, which accounts for relative URLs where `new URL()` does not

Reply:
But it just throws an error

Reply:
And this needs to work in node in order to render the initialUrl